haicrypt/cryspr-openssl-evp.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * SRT - Secure, Reliable, Transport
+ * Copyright (c) 2019 Haivision Systems Inc.
+ * 
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * 
+ */
+
+
+/*****************************************************************************
+written by
+   Haivision Systems Inc.
+
+   2019-06-26 (jdube)
+        OpenSSL CRYSPR/4SRT (CRYypto Service PRovider for SRT).
+*****************************************************************************/
+
+#include "hcrypt.h"
+
+#include <string.h>
+
+
+typedef struct tag_crysprOpenSSL_EVP_cb {
+        CRYSPR_cb       ccb;
+        /* Add cryptolib specific data here */
+} crysprOpenSSL_EVP_cb;
+
+
+int crysprOpenSSL_EVP_Prng(unsigned char *rn, int len)
+{
+    return(RAND_bytes(rn, len) <= 0 ? -1 : 0);
+}
+
+const EVP_CIPHER* (*Xcipher_fnptr)(void) = EVP_aes_128_ecb;
+
+const EVP_CIPHER* (*_crysprOpenSSL_EVP_cipher_fnptr[][3])(void) = {
+    {NULL,NULL,NULL},
+    {EVP_aes_128_ecb, EVP_aes_192_ecb, EVP_aes_256_ecb},
+    {EVP_aes_128_ctr, EVP_aes_192_ctr, EVP_aes_256_ctr},
+};
+
+
+int crysprOpenSSL_EVP_AES_SetKey(
+    int cipher_type,            /* One of HCRYPT_CTX_MODE_[CLRTXT|AESECB|AESCTR|AESGDM] */
+    bool bEncrypt,              /* true Enxcrypt key, false: decrypt */
+    const unsigned char *kstr,  /* key sttring*/
+    size_t kstr_len,            /* kstr len in  bytes (16, 24, or 32 bytes (for AES128, AES192, or AES256) */
+    CRYSPR_AESCTX *aes_key)     /* CRYpto Service PRovider AES Key context */
+{
+    const EVP_CIPHER *cipher = NULL;
+    int idxKlen = (kstr_len/8)-2; /* key_len index in cipher_fnptr array in [0,1,2] range */
+
+    switch(cipher_type) {
+    case HCRYPT_CTX_MODE_CLRTXT:
+        return 0;
+    case HCRYPT_CTX_MODE_AESECB:
+        break;
+    case HCRYPT_CTX_MODE_AESCTR:
+#if !CRYSPR_HAS_AESCTR
+        /* internal implementation of AES-CTR using crypto lib's AES-ECB */
+        cipher_type = HCRYPT_CTX_MODE_AESECB;
+#endif
+        break;
+    default:
+        HCRYPT_LOG(LOG_ERR, "invalid cipher type (%d). Expected: [%d..%d]\n",
+                   cipher_type, HCRYPT_CTX_MODE_AESECB, HCRYPT_CTX_MODE_AESCTR);
+        return(-1);
+    }
+    switch(kstr_len) {
+    case 128/8:
+    case 192/8:
+    case 256/8:
+        break;
+    default:
+        HCRYPT_LOG(LOG_ERR, "invalid key length (%d). Expected: 16, 24, 32\n", (int)kstr_len);
+        return -1;
+    }
+    cipher = _crysprOpenSSL_EVP_cipher_fnptr[cipher_type][idxKlen]();
+
+    if (bEncrypt) {        /* Encrypt key */
+        if(!EVP_CipherInit_ex(aes_key, cipher, NULL, kstr, NULL, 1)) {
+            HCRYPT_LOG(LOG_ERR, "%s", "EVP_CipherInit_ex(kek) failed\n");
+            return(-1);
+        }
+    } else {               /* Decrypt key */
+        if(!EVP_CipherInit_ex(aes_key, cipher, NULL, kstr, NULL, 0)) {
```

Comment:
Also possible using `EVP_EncryptInit_ex` and `EVP_DecryptInit_ex`:

```suggestion
    if (bEncrypt) {        /* Encrypt key */
        if(!EVP_EncryptInit_ex(aes_key, cipher, NULL, kstr, NULL)) {
            HCRYPT_LOG(LOG_ERR, "%s", "EVP_CipherInit_ex(kek) failed\n");
            return(-1);
        }
    } else {               /* Decrypt key */
        if(!EVP_DecryptInit_ex(aes_key, cipher, NULL, kstr, NULL)) {
```

Otherwise can be just:
```c++
if(!EVP_CipherInit_ex(aes_key, cipher, NULL, kstr, NULL, (int) bEncrypt)) {
```

Reply:
1st suggestio implemented